### ISSUE TITLE
When using sphinxcontrib.bibtex, rinohtype doesn't render citations

### ISSUE THREAD

**Issue by rappdw** [brechtm/rinohtype#298]
I've distilled the project down to bare minimum here: https://github.com/rappdw/rinoh_error

`make html` produces expected results
`make rinoh` results in error

rinohtype version: 0.5.3 or git+https://github.com/brechtm/rinohtype.git@master
sphinx version: 4.2.0
python version: 3.9.7
os: MacOS 11.6

The log file is in the repository, but I've also included it here:

```
# Sphinx version: 4.2.0
# Python version: 3.9.7 (CPython)
# Docutils version: 0.17.1 release
# Jinja2 version: 3.0.1
# Last messages:
#   building [rinoh]: all documents
#   updating environment:
#   0 added, 0 changed, 0 removed
#   looking for now-outdated files...
#   none found
#   processing book...
#   index
#   chapter_1
#   bibliography
#   resolving references...
# Loaded extensions:
#   sphinx.ext.mathjax (4.2.0) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.2) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.2) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (2.0.0) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.5) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.3) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/alabaster/__init__.py
#   rinoh.frontend.sphinx (0.5.3) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/sphinx/__init__.py
#   myst_parser (0.15.2) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/myst_parser/__init__.py
#   sphinxcontrib.bibtex (2.4.1) from /Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinxcontrib/bibtex/__init__.py
Traceback (most recent call last):
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinx/cmd/build.py", line 280, in build_main
    app.build(args.force_all, filenames)
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinx/application.py", line 343, in build
    self.builder.build_update()
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 290, in build_update
    self.build(['__all__'], to_build)
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/sphinx/builders/__init__.py", line 357, in build
    self.write(docnames, list(updated_docnames), method)
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/sphinx/__init__.py", line 246, in write
    self.write_document(entry)
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/sphinx/__init__.py", line 252, in write_document
    rinoh_document = self.construct_rinohtype_document(data)
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/sphinx/__init__.py", line 268, in construct_rinohtype_document
    rinoh_tree = from_doctree(doctree, sphinx_builder=self)
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 141, in from_doctree
    return mapped_tree.flowable()
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 132, in flowable
    flowable, = self.flowables()
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 99, in flowables
    for flowable in super().flowables():
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 137, in flowables
    for i, flowable in enumerate(self.build_flowables()):
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 146, in build_flowables
    yield self.build_flowable()
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 162, in build_flowable
    return self.grouped_flowables_class(self.children_flowables(),
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 125, in children_flowables
    return list(chain(*(item.flowables() for item in children)))
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 99, in flowables
    for flowable in super().flowables():
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 137, in flowables
    for i, flowable in enumerate(self.build_flowables()):
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 146, in build_flowables
    yield self.build_flowable()
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 162, in build_flowable
    return self.grouped_flowables_class(self.children_flowables(),
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 125, in children_flowables
    return list(chain(*(item.flowables() for item in children)))
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 99, in flowables
    for flowable in super().flowables():
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 137, in flowables
    for i, flowable in enumerate(self.build_flowables()):
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 146, in build_flowables
    yield self.build_flowable()
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/sphinx/nodes.py", line 232, in build_flowable
    return super().build_flowable(id='%' + self.get('docname'), **kwargs)
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 162, in build_flowable
    return self.grouped_flowables_class(self.children_flowables(),
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 125, in children_flowables
    return list(chain(*(item.flowables() for item in children)))
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 99, in flowables
    for flowable in super().flowables():
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 137, in flowables
    for i, flowable in enumerate(self.build_flowables()):
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 146, in build_flowables
    yield self.build_flowable()
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 162, in build_flowable
    return self.grouped_flowables_class(self.children_flowables(),
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 125, in children_flowables
    return list(chain(*(item.flowables() for item in children)))
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 99, in flowables
    for flowable in super().flowables():
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 137, in flowables
    for i, flowable in enumerate(self.build_flowables()):
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 146, in build_flowables
    yield self.build_flowable()
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/nodes.py", line 208, in build_flowable
    return rt.Paragraph(self.process_content(), style=self.style)
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 78, in process_content
    return MixedStyledText([text for text in children_text if text],
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 78, in <listcomp>
    return MixedStyledText([text for text in children_text if text],
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/rst/__init__.py", line 77, in <genexpr>
    children_text = (child.styled_text() for child in self.getchildren())
  File "/Users/drapp/opt/miniconda3/envs/rinoh_error/lib/python3.9/site-packages/rinoh/frontend/__init__.py", line 42, in __getattr__
    raise AttributeError('No such element: {} in {}'.format(name, self))
AttributeError: No such element: styled_text in <rinoh.frontend.rst.nodes.Citation object at 0x7f93f8adfb20>
```

**Comment by brechtm** [comment]
Thank you for the repository reproducing the issue.

I think this is a bug in sphinxcontrib-bibtex. The document tree it produces is:

```xml
<document source="/Users/brechtm/Code/rt_debug/298_bibtex_citation/source/bibliography.md">
    <section classes="tex2jax_ignore mathjax_ignore" ids="bibliography" myst-anchor="bibliography.md#bibliography" names="bibliography">
        <title>
            Bibliography
        <paragraph ids="id1">
            <citation docname="bibliography" ids="id2">
                <label support_smartquotes="False">
                    1
                <paragraph>
                    Donald ...
```

The _citation_ node is a child of a _paragraph_, while the docutils specification lists _citation_ as a [body element](https://docutils.sourceforge.io/docs/ref/doctree.html#body-elements), **not** an inline element. To fix this, the wrapping _paragraph_ should be eliminated. For reference, look at the output of `rst2pseudoxml.py` (included with _docutils_) on a [citation defined in plain reStructured](https://docutils.sourceforge.io/docs/ref/rst/restructuredtext.html#citations). For example:

```rst
Here is a citation reference: [CIT2002]_.

.. [CIT2002] This is the citation.  It's just like a footnote,
   except the label is textual.
```

```xml
<document source="/tmp/bleh.rst">
    <paragraph>
        Here is a citation reference:
        <citation_reference ids="id1" refid="cit2002">
            CIT2002
        .
    <citation backrefs="id1" ids="cit2002" names="cit2002">
        <label>
            CIT2002
        <paragraph>
            This is the citation.  It's just like a footnote,
            except the label is textual.
```

I have noticed before that the builder included with Sphinx don't make a fuss out of this. I think it makes sense to make extensions respect the docutils doctree specs though. Please report this bug at the sphinxcontrib-bibtex issue tracker.

**Comment by rappdw** [comment]
As per comments on https://github.com/mcmtroffaes/sphinxcontrib-bibtex/pull/274#issuecomment-934239780

With sphinxcontrib-bibtex now conforming to the docutils spec, a pdf file is now generated, but it is missing the bibliography citations. 

Should this issue be reopened, or should I create a separate issue?

**Comment by brechtm** [comment]
I discovered that rinohtype didn't support the case where a citation reference and the corresponding citation are not located in the same file. After fixing that (yet to be committed), I was able to get some output from the sample Sphinx project provided.

Unfortunately, the citation's label fails to render properly. The citation reference produced by sphinxcontrib-bibtex is a plain _reference_ as opposed to a _citation_reference_ node and rinohtype's citation handling trips over this. I'll report this on sphinxcontrib-bibtex's issue tracker to see whether this could be changed.
